### PR TITLE
Add JRM operator and associated tests

### PR DESCRIPTION
**Pull request by ziyiyin97**
1. Add joJRM
2. add associated tests
3. overload multiplication of jolinearfunction with vector of abstract vectors so JRM works not only for vector of vectors, but also for vector of judiVectors/modelParameters/etc in JUDI

Appreciate any comment

### CODE REVIEW

File: src/joLinearFunctionConstructors/joJRM.jl [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# joJRM operator
+
+## helper module
+module joJRM_etc
+
+    function apply_fwd(n, L, z; γ=1f0)
+        return view(z, n+1:length(z)) .+ 1f0/γ * repeat(view(z, 1:n), L)    # get original components
```
Comment (mloubout) [inline src/joLinearFunctionConstructors/joJRM.jl:7]: view don't gain you anything you are creating and returning a new array.

File: src/joLinearFunctionConstructors/joJRM.jl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    end 
+
+    function apply_adj(L, x; γ=1f0)
+        x0 = 1f0/γ * sum(reshape(view(x,:), :, L), dims=2)   # weighted sum on common component
```
Comment (mloubout) [inline src/joLinearFunctionConstructors/joJRM.jl:11]: probably don't gain much neither with view here.

File: src/joLinearFunctionConstructors/joJRM.jl [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        return vcat(x0, x)
+    end
+
+end
+using .joJRM_etc
+
+export joJRM
+"""
+    julia> op = joJRM(ops;[name=...])
```
Comment (mloubout) [inline src/joLinearFunctionConstructors/joJRM.jl:20]: This is not very descriptive, a generic user will have zero idea what this is.

And again, JRM is the inversion method, this is just a block operator not an algorithm.  It's not said anywhere what JRM even means

File: src/joLinearFunctionConstructors/joJRM.jl [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+# Signature
+
+    joJRM(ops::Vector{joAbstractLinearOperator}; γ::Number=1f0, name::String="joJRM")
+
+# Arguments
+
+- `ops`: an array of JOLI operators (subtypes of joAbstractLinearOperator)
+- keywords
+    - `γ`: weight of common component = 1/γ "Xiaowei Li, A weighted ℓ1-minimization for distributed compressive sensing"
```
Comment (mloubout) [inline src/joLinearFunctionConstructors/joJRM.jl:32]: common component is in `z` not in the operator need better desripton.

File: src/joLinearFunctionConstructors/joJRM.jl [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        ops[i].n==opn || throw(joLinearFunctionException("domain length mismatch"))
+    end
+    As = joCoreBlock(ops...)
+    Φ = joLinearFunctionFwd(L*ops[1].n, (L+1)*ops[1].n,
+        v1 -> joJRM_etc.apply_fwd(ops[1].n, L, v1; γ=γ),
+        v2 -> joJRM_etc.apply_adj(L, v2; γ=γ),
+        v3 -> joJRM_etc.apply_adj(L, v3; γ=γ),
+        v4 -> joJRM_etc.apply_fwd(ops[1].n, L, v4; γ=γ),
+        DDT, RDT; name=name)
+    return As * Φ
```
Comment (mloubout) [inline src/joLinearFunctionConstructors/joJRM.jl:72]: That's not great. this returns a generic `joLinearOperator`, and the "name" is hidden in a sub-operator to dig for. I find this setup quite confusing and not very user friendly.

File: test/test_joJRM.jl [hunk 6]
```diff
@@ -0,0 +1,17 @@
+T=3
```
Comment (mloubout) [inline test/test_joJRM.jl:1]: Why is this now called `T` was `L`  in the operator definition.

File: test/test_joJRM.jl [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+tsname="joJRM"
+@testset "$tsname" begin
+    for t=2:T # start test loop
+    
+        ops = [joMatrix(randn(Float32, 2*t+8, 2*t)) for i = 1:t]
+        γ = abs(randn(Float32))
+        A=joJRM(ops; γ=γ)
+
+        verbose && println("$tsname ($m)")
+        @testset "1 common component and $t innovation components" begin
```
Comment (mloubout) [inline test/test_joJRM.jl:11]: That;s what you apply the operator to, no what the operator is.

**Review by mloubout — CHANGES_REQUESTED** [review]
While I understand what this does and aim to provide because I am familiar with the method and work that goes with it, I think this is not very user friendly nor generic in a way that belongs in JUDI, this should be in you own JRM package or made generic. But the generic form is just a `joCoreBlock`  so I]m not sure what this adds.

**Comment by ziyiyin97** [comment]
Thanks for your comments. I will close this PR and restructure this into Seis4CCS

**Pull request closed without merging**